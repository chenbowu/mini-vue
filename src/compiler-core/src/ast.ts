export enum NodeTypes {
  INTERPOLATION,
  ELEMENT,
  SIMPLE_EXPRESSION,
}
