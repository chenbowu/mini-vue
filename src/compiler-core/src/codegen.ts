import { NodeTypes } from './ast'
import { CREATE_ELEMENT_VNODE, TO_DISPLAY_STRING, helperMapName } from './runtimeHelpers'

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  console.log('codegen--------', ast)
  if (ast.helpers.length > 0)
    genFunctionPreamble(ast, context)

  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')

  push('return ')
  push(`function ${functionName}(${signature}) {`)
  push('return ')
  genNode(ast.codegenNode, context)
  push('}')
  return {
    code: context.code,
  }
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    default:
      break
  }
}

function createCodegenContext() {
  const context = {
    code: '',
    push: (code) => {
      context.code += code
    },
    helper: key => `_${helperMapName[key]}`,
  }
  return context
}

function genFunctionPreamble(ast, context) {
  const { push } = context
  const VueBinging = 'Vue'
  const aliasHelper = (s: any) => `${s}: _${s}`
  push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${VueBinging}`)
  push('\n')
}

function genText(node: any, context: any) {
  const { push } = context
  push(`"${node.content}"`)
}

function genInterpolation(node: any, context: any) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

function genExpression(node: any, context: any) {
  const { push } = context
  push(node.content)
}

function genElement(node, context) {
  const { push, helper } = context
  push(`_${helper(CREATE_ELEMENT_VNODE)}("div", null, "hi," + _toDisplayString(_ctx.message), 1)`)
}
