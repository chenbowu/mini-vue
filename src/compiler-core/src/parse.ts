import { NodeTypes } from './ast'

const enum TagType {
  Start,
  End,
}

export function baseParse(content: string) {
  const context: ParserContext = createParserContext(content)
  return createRoot(parseChildren(context, []))
}

function parseChildren(context: ParserContext, ancestors: any[]) {
  const nodes: any = []
  let node
  while (!isEnd(context, ancestors)) {
    if (context.source.startsWith('{{')) {
      node = parseInterpolation(context)
    }
    else if (context.source.startsWith('<')) {
      if (/[a-z]/i.test(context.source[1]))
        node = parseElement(context, ancestors)
    }
    else {
      node = parseText(context)
    }

    nodes.push(node)
  }
  return nodes
}

function isEnd(context: ParserContext, ancestors: any[]) {
  const s = context.source
  if (s.startsWith('</')) {
    for (let i = 0; i < ancestors.length; i++) {
      const tag = ancestors[i].tag
      if (startsWithEndTagOpen(context, tag))
        return true
    }
  }
  return !s
}

function startsWithEndTagOpen(context, tag: string) {
  return (context.source.startsWith('</') && context.source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase())
}

function parseText(context: ParserContext) {
  // 限定文本截取范围默认为所有
  let endIndex = context.source.length
  // 文本截取结束标识
  const endToken = ['</', '{{']
  // 判断文本中是否存在 token 标识，存在则文本截取到此 token 为止, 否则截取到文本末尾.
  for (let i = 0; i < endToken.length; i++) {
    const index = context.source.indexOf(endToken[i])
    if (index !== -1 && index < endIndex)
      endIndex = index
  }
  const content = parseTextData(context, endIndex)
  return {
    type: NodeTypes.TEXT,
    content,
  }
}

/**
 * Get text data with given length from the current location
 * @param context
 * @param length
 * @returns
 */
function parseTextData(context: ParserContext, length: number): string {
  const rawText = context.source.substring(0, length)
  advanceBy(context, length)
  return rawText
}

function parseElement(context: ParserContext, ancestors: any[]) {
  const element: any = parseTag(context, TagType.Start)
  ancestors.push(element)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()
  const tag = element.tag

  if (startsWithEndTagOpen(context, tag))
    parseTag(context, TagType.End)
  else
    throw new Error(`缺少结束标签: ${tag}`)

  return element
}

function parseTag(context: ParserContext, type: TagType) {
  const match = /^<\/?([a-z]*)/i.exec(context.source)
  const tag = match![1]
  advanceBy(context, match![0].length)
  advanceBy(context, 1)

  if (type === TagType.End)
    return

  return {
    type: NodeTypes.ELEMENT,
    tag,
  }
}

function parseInterpolation(context: ParserContext) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiter.length)
  advanceBy(context, openDelimiter.length)
  const rawContentLength = closeIndex - openDelimiter.length
  const rawContent = parseTextData(context, rawContentLength)
  const content = rawContent.trim()
  advanceBy(context, closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  }
}

function advanceBy(context, step) {
  context.source = context.source.slice(step)
}

function createParserContext(content: string): ParserContext {
  return {
    source: content,
  }
}

function createRoot(children) {
  return {
    children,
  }
}

interface ParserContext {
  source: string
}

