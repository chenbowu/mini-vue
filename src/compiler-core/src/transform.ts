import { NodeTypes } from './ast'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  traverseNode(root, context)
  createRootCodegen(root)

  root.helpers = [...context.helpers.keys()]
}

function traverseNode(node, context) {
  const { nodeTransforms } = context
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    transform(node, context)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context)
      break
    default:
      break
  }
}

function traverseChildren(node, context) {
  const children = node.children
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traverseNode(node, context)
    }
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    },
  }
  return context
}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0]
}

