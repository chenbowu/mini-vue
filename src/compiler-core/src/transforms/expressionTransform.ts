import { NodeTypes } from '../ast'

export function expressionTransform(node) {
  if (node.type === NodeTypes.INTERPOLATION)
    node.content = processExpression(node.content)
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`
  return node
}
