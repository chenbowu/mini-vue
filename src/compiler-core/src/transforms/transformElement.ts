import { NodeTypes } from '../ast'
import { CREATE_ELEMENT_VNODE, helperMapName } from '../runtimeHelpers'

export function transformElement(node, context) {
  return () => {
    if (node.type === NodeTypes.ELEMENT) {
      context.helper(helperMapName[CREATE_ELEMENT_VNODE])
      const { tag, props, children } = node

      // TODO ROOT 和 ELEMENT 需要创建一个 codegenNode, 入口节点
      // 将所有子节点封装 codegenNode 中，对次节点类型进行特殊处理
      // 使其 createVNode 时可传入 vnode 数组
      const vnodeTag = `"${tag}"`
      const vnodeProps = props
      const vnodeChildren = children[0]

      const vnodeElement = {
        type: NodeTypes.ELEMENT,
        tag: vnodeTag,
        props: vnodeProps,
        children: vnodeChildren,
      }

      node.codegenNode = vnodeElement
    }
  }
}
