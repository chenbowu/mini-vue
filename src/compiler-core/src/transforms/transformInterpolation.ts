import { NodeTypes } from '../ast'
import { TO_DISPLAY_STRING, helperMapName } from '../runtimeHelpers'

export function transformInterpolation(node, context) {
  if (node.type === NodeTypes.INTERPOLATION)
    context.helper(helperMapName[TO_DISPLAY_STRING])
}
