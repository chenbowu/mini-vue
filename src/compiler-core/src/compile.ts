import { generate } from './codegen'
import { baseParse } from './parse'
import { transform } from './transform'
import { transformElement } from './transforms/transformElement'
import { transformExpression } from './transforms/transformExpression'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'

export function baseCompile(template) {
  const ast = baseParse(template)
  transform(ast, {
    nodeTransforms: [transformExpression, transformInterpolation, transformElement, transformText],
  })

  return generate(ast)
}
