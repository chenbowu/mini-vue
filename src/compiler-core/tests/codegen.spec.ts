import { generate } from '../src/codegen'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'
import { transformElement } from '../src/transforms/transformElement'
import { transformExpression } from '../src/transforms/transformExpression'
import { transformInterpolation } from '../src/transforms/transformInterpolation'

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}')
    transform(ast, {
      nodeTransforms: [transformInterpolation, transformExpression],
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('compound', () => {
    const ast = baseParse('<div>hi,{{message}}</div>')
    transform(ast, {
      nodeTransforms: [transformExpression, transformInterpolation, transformElement],
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
