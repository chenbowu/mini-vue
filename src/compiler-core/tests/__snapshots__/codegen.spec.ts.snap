// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen compound 1`] = `
"const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {return __createElementVNode(\\"div\\", null, \\"hi,\\" + _toDisplayString(_ctx.message), 1)}"
`;

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache) {return \\"hi\\"}"`;
